tp: allow the HTTP RPC accelerator to open more than 1 file

WangJialin547 opened the pull request:
Added a feature that solves the issue that at most one tab can be using the http + rpc accelerator. 

Revised #2597 and made the change focused and minimal

WangJialin547 committed:
feat: allow multiple tabs to use the HTTP RPC accelerator

WangJialin547 committed:
attempt to solve problem 1

LalitMaganti commented:
I had a looked a the proposed fix and I'm reasonably sure it's not going to work if the lid is closed for a few hours (the laptop will wake up every now and again but likely will not connect within the timeout). 

Sketching out how I expect a solution to work like:
1) when the websocket is created, the first message the UI sends is a new message indicating the UUID it wants to use for the trace processor associated with the websocket 
2) rpc keeps a map from uuid -> trace processor and uses this to figure out what trace processor to use for connection. If no uuid is given (for back compat) it uses the "global" trace processor as it does today. 
3) these trace processors are pruned after a long timeout (O(hours) without activity or number of messages etc). There should be an option to set the timeout from the command line
4) to solve problem 2, I said above, we need a new dialog in UI (with an expansion of the /status response) where the rpc provides all the trace processor instances in the UUID map and allows the user to pick which one they want to connect to (this handled both the case of timeouts, accidentally closing browser tab and also preloading the trace from command line and using that from the UI). 

So TLDR the solution has to be significantly more complex than just changing the rpc layer: it will need changes to the UI as well: this is unavoidable. 

@sashwinbalaji fyi.

WangJialin547 committed:
attempt to solve problem 1

WangJialin547 committed:
added thread pruning mechanic

WangJialin547 committed:
Merge branch 'problem1'

WangJialin547 committed:
temp save

WangJialin547 committed:
changed proto, added modal

WangJialin547 committed:
temp save

WangJialin547 committed:
format code

WangJialin547 committed:
fix modal 2

WangJialin547 committed:
solved problem 2

WangJialin547 committed:
format files

WangJialin547 committed:
fixed missing file name bug

WangJialin547 committed:
added missing message handling

WangJialin547 committed:
fixed segfault issue

WangJialin547 commented:
@LalitMaganti @sashwinbalaji 

both issues 1 and 2 have been addressed according to the instructions. this current modification allows the user to open multiple files using the trace processor as the backend establishes multiple trace processors according to the connections, and the "trace processor is already being used" modal is updated so the user can choose the desired file. 

However there is a limitation that each trace processor remains limited to only one tab at a time. in order for a single trace processor to be accessible for multiple tabs i need to remove the rpc sequence mismatch check and update the restoreInitialTables so the TP doesn't clear tables when it restarts and both tabs can access their respctive tabes, but i'm not sure if that's allowed

WangJialin547 committed:
update comments

LalitMaganti commented:
> However there is a limitation that each trace processor remains limited to only one tab at a time. in order for a single trace processor to be accessible for multiple tabs i need to remove the rpc sequence mismatch check and update the restoreInitialTables so the TP doesn't clear tables when it restarts and both tabs can access their respctive tabes, but i'm not sure if that's allowed

We definitely *don't* want to support that the same TP instance is accessible from multiple tabs at the same time. So we should keep the sequence checks and the restore initial state. 

@sashwinbalaji will take care of reviewing this.

sashwinbalaji commented on src/trace_processor/rpc/httpd.cc:
```diff
@@ -262,13 +426,120 @@ void Httpd::OnHttpRequest(const base::HttpRequest& req) {
+    auto it = conn_to_uuid_map.find(msg.conn);
+    // if we cannot find the connection and uuid being registered then the ui is
+    // older. we fall back to the global rpc for backward compatibility
+    if (it == conn_to_uuid_map.end() || it->second == "default") {
```
make "default" a constant and use that in OnWebsocketMessage and OnHttpRequest

sashwinbalaji commented on src/trace_processor/rpc/httpd.cc:
```diff
@@ -149,11 +269,55 @@ void Httpd::OnHttpRequest(const base::HttpRequest& req) {
+    }
+
+    // Add per-UUID trace processors
+    for (const auto& entry : uuid_to_tp_map) {
```
this access is not protected by websocket_rpc_mutex_ ?

sashwinbalaji commented on src/trace_processor/rpc/httpd.cc:
```diff
@@ -117,6 +214,29 @@ void Httpd::Run(const std::string& listen_ip,
+    cleanUpInactiveInstances();
+    if (tp_timeout_mins_ > 0) {
+      task_runner_.PostDelayedTask(
+          *cleanup_task, static_cast<uint32_t>(tp_timeout_mins_ * 60 * 1000));
```
nit: make it constexpr like kMillisecondPerMinute and reuse

sashwinbalaji commented on ui/src/frontend/rpc_http_dialog.ts:
```diff
@@ -155,23 +156,33 @@ export async function CheckHttpRpcConnection(): Promise<void> {
+  const tpStatusAll = assertExists(state.status);
+
+  // use the first trace processor if available, otherwise fallback
+  const firstStatusData = tpStatusAll.traceProcessorStatuses?.[0]?.status;
```
nit: firstTpStatusData

WangJialin547 committed:
fix typo

WangJialin547 committed:
remove unneccessary timeout_mins parameter

WangJialin547 committed:
changed UuidRpcThread to a class

WangJialin547 committed:
renamed IsUuidHandshake function to PascalCase

WangJialin547 committed:
replaced magic numbers with constexprs kMillisecondPerMinute and kNanosecondPerMin

WangJialin547 committed:
renamed firstStatus into firstTpStatus

WangJialin547 committed:
modify sendTraceTitle method

WangJialin547 committed:
added mutex protection for /status/all endpoint

WangJialin547 committed:
changed default string into a constant

WangJialin547 committed:
changed it and it2 to descriptive var name

WangJialin547 committed:
removed unneccessary size 5 check

WangJialin547 committed:
changed while loop in cleanUpInactiveInstances to for loop

WangJialin547 committed:
added uuid local variable in the loop inside of cleanUpInactiveInetances

WangJialin547 committed:
fix

WangJialin547 committed:
changed atoi to base::StringToInt32

WangJialin547 committed:
removed wrapper function isUuidHandshake

WangJialin547 committed:
fixed comment

WangJialin547 committed:
modified comments

sashwinbalaji reviewed (changes requested):
Thanks for your contribution.

So I tested it out locally and found some bugs/improvements:

1)  If I go to use loaded trace and then cancel, crash happens.

![popup_crash](https://github.com/user-attachments/assets/0e9dfa36-09c4-40c1-895f-9c1fe2c6cf1d)

2) After adding support for multipl tps, we should just use it instead of assignig "default" for the first loaded trace. This gives a confusing ux. 
<img width="2904" height="1230" alt="4zskuqsynAEgRhy" src="https://github.com/user-attachments/assets/84e1b511-a3f1-4726-8bdf-36d42df2ee2a" />

3) The behavior of letting user open a already loaded trace in a new tab without any prior notification and killing it in the old one is not a user friendly experience as in the older tab they might have some modifications. Ideally we should show all the already used traces at the bottom with explicitly telling the user if they want to use that trace they are going to kill it on the older one.

4) the two popup flow is not the user friendly. Let's merge them into one. Something like below:

```
Use trace processor native acceleration?

Trace Processor detected on localhost:9001 one or more loaded traces including
/Users/sashwinbalaji/Desktop/my_custom_trace.pftrace (0 MB).

YES, select loaded trace:

List down all the loaded traces (in order of not being used followed by used).
For used either add a heading above saying used / add a button for each of those lines saying override.

YES, but reset state:
Use this if you want to open another trace but still use the
accelerator.

NO, Use builtin WASM:
Will not use the accelerator in this tab.

Using the native accelerator has some minor caveats:
- Sharing, downloading and conversion-to-legacy aren't supported.
- Each trace file can be opened in at most one tab at a time.
```

Looping in @stevegolton to take a look into it from UI side and provide feedback.

WangJialin547 committed:
updated build.gn for httpd to include ../../base:version

WangJialin547 committed:
supports parsing uuid from message using json_utils.h

WangJialin547 committed:
fix

WangJialin547 committed:
addressed issue 2: the preloaded trace using trace_processor_shell gets is own uuid and thread instead of using the global rpc

WangJialin547 committed:
added a hasActiveTab field to the status message

WangJialin547 committed:
new version of popup modal

WangJialin547 committed:
made processors with active tabs accessible, but display warning

WangJialin547 committed:
addressed problems 1, 3 and 4 by introducing a new modal

WangJialin547 committed:
fixed race condition that cuases a segfault

WangJialin547 committed:
moved the buttons to the bottom like the old version of the dialog

WangJialin547 committed:
format code

WangJialin547 committed:
fix

WangJialin547 commented:
@sashwinbalaji 

All four issues have been addressed.

Issue 2 was addressed by updating the constructor in httpd.cc, and issues 1, 3, and 4 have been addressed by introducing a brand-new, unified modal to allow the user to choose the desired file with all the processors that have an active tab at the bottom, displaying a warning that the user should close the old tab before opening the new one. Clicking cancel doesn't cause a crash anymore.

WangJialin547 committed:
fixed warning message

WangJialin547 committed:
added a mouse hover effect on the file selection buttons

WangJialin547 committed:
fixed bug: now trace_processor_shell with no preloaded trace won't create an empty thread anymore

WangJialin547 committed:
attempt to solve CI failure by including json folder in build.gn

WangJialin547 committed:
update comment

stevegolton commented on ui/src/frontend/index.ts:
```diff
@@ -124,6 +124,8 @@ function setupContentSecurityPolicy() {
       'ws://127.0.0.1:8037', // For the adb websocket server.
       'https://*.google-analytics.com',
       'https://*.googleapis.com', // For Google Cloud Storage fetches.
+      `ws://${window.location.hostname}:9001`,
```
A couple of things:
- Why do we need this? It's unlikely that a trace processor instance will be hosted on ui.perfetto.dev for instance.
- They should probably go into the `rpcPolicy` list above anyway.

sashwinbalaji commented:
@WangJialin547 , thanks for all the effort you've put into this! The team is going to discuss in more detail internally to finalize the long term plan for this critical feature. To avoid any unnecessary work on your end, could you please pause on any more changes for now? We'll get back to you as soon as we have a clear path forward.

primiano commented:
[Perfetto UI & Multi trace-processor instances.pdf](https://github.com/user-attachments/files/22095940/Perfetto.UI.Multi.trace-processor.instances.pdf)

primiano commented:
Hey sorry for all the back and forth here. We had some internal discussion as this is something we wanted to do for a while and few of us (including myself) has some opinions on how this should be done.

I put together a writeup in this document, attached here:
[Perfetto UI & Multi trace-processor instances.pdf](https://github.com/user-attachments/files/22095961/Perfetto.UI.Multi.trace-processor.instances.pdf)

I'm not asking to do 100% what's written there, but anything that goes in that direction would be more than welcome.

WangJialin547 committed:
fixed proto file

WangJialin547 committed:
shows the trace processor selection modal in all situations

WangJialin547 committed:
removed /status/all endpoint

WangJialin547 committed:
when websocket disconnects, removes tp instance if it has no trace loaded

WangJialin547 committed:
fixed the problem where tools/format-sources is complaining

WangJialin547 committed:
apply changes to the modal

WangJialin547 committed:
fixed rpc_http_dialog

WangJialin547 committed:
added new endpoints for websocket

WangJialin547 committed:
solved the race condition problem when closing laptop lid by adding a really short timeout

WangJialin547 committed:
cleanup work, removing unused functions and clarifying comments

WangJialin547 committed:
changed uuid to monotonic id for tps

WangJialin547 committed:
fixed eslint issues when formatting

WangJialin547 committed:
fixed presubmit issues

WangJialin547 committed:
fixed version mismatch error

WangJialin547 committed:
fixed backcomp issue by sending the uuid wrapped in a protobuf message with statusResult

WangJialin547 committed:
modified modal to fit the requirements

WangJialin547 committed:
moved rpcPolicy

WangJialin547 committed:
moved inline styles to a style sheet

WangJialin547 committed:
removed unneccessary include

WangJialin547 committed:
'New Instance...' won't show if there are no tp instances detected as it is the default behavior

WangJialin547 committed:
format code

WangJialin547 committed:
fix presubmit issue

WangJialin547 commented:
Hi @sashwinbalaji ,

I have adjusted the code as requested. currently the modal looks like this: 

<img width="2560" height="1528" alt="Screenshot 2025-09-13 195743" src="https://github.com/user-attachments/assets/808c47f1-fd5b-45fc-ad08-11277a07732b" />

Which includes a list of all existing instances that allows the user to select from. In addition, the user can select "New Instance...", which will create a new tp instance and connect to it via ws. 

Also I noticed that sending a plain json ws message will cause a rpc framing error in older UI, so I wrapped it inside an rpc TPM_GET_STATUS message which will pass the proto ring buffer tag check. 

Please let me know if I should make any changes

WangJialin547 commented:
<img width="2560" height="1528" alt="image" src="https://github.com/user-attachments/assets/2eb3d565-b8c4-437f-9e6f-f9f3e68cc669" />
<img width="2560" height="1528" alt="image" src="https://github.com/user-attachments/assets/bd4ed7fb-fc34-4293-9853-18e52b45df48" />

sashwinbalaji commented:
on trying to open multiple traces, trace_processor_shell crashes, please fix this.

Also though the ui seems better aligned with the doc, it still doesn't give the option to user to close an unused session.
<img width="976" height="389" alt="image" src="https://github.com/user-attachments/assets/0ae93821-c7b9-4e42-9db3-c1d53bc681b4" />

nit: it would be better if the unused sessions are sorted in ascending order from top to down in the UI.

steps to reproduce the crash:

- Run: tp shell (out/linux_clang_debug/trace_processor_shell --httpd) and ui(ui/run-dev-server) locally
- open two traces on separate tab
- close the second tab.

```
[776.987]       http_server.cc:67 [HTTP] Starting HTTP server on 127.0.0.1:9001
[776.987]       http_server.cc:78 [HTTP] Starting HTTP server on [::1]:9001
[776.987]            httpd.cc:262 [HTTP] This server can be used by reloading https://ui.perfetto.dev and clicking on YES on the "Trace Processor native acceleration" dialog or through the Python API (see https://perfetto.dev/docs/analysis/trace-processor#python-api).
[776.987]            httpd.cc:265 RPC timeout enabled: 120 minutes
[856.426]      http_server.cc:120 [HTTP] New connection
[856.427]      http_server.cc:271 [HTTP] POST /status [body=0B, origin="http://localhost:10000"]
[856.439]      http_server.cc:271 [HTTP] POST /status [body=0B, origin="http://localhost:10000"]
[874.581]      http_server.cc:271 [HTTP] POST /status [body=0B, origin="http://localhost:10000"]
[874.589]      http_server.cc:120 [HTTP] New connection
[874.589]      http_server.cc:271 [HTTP] GET /websocket/new [body=0B, origin="http://localhost:10000"]
[874.589]            httpd.cc:420 New TP instance 1 created via /websocket/new
[874.589]      http_server.cc:358 [HTTP] Handshaking WebSocket for /websocket/new
[874.589]      http_server.cc:360 > Upgrade: websocket
[874.589]      http_server.cc:360 > Connection: Upgrade
[874.589]      http_server.cc:360 > Sec-WebSocket-Accept: JWkU4wK3unGLyIoHttiK/QW60K8=
[874.647] ing_trace_parser.cc:108 proto trace detected
[874.802]    proto_decoder.cc:140 Skipping field 102672883 because its id > 16777215
[874.803]    proto_decoder.cc:140 Skipping field 102672883 because its id > 16777215
[874.804]    proto_decoder.cc:140 Skipping field 102672883 because its id > 16777215
[874.804]    proto_decoder.cc:140 Skipping field 102672883 because its id > 16777215
[874.804]    proto_decoder.cc:140 Skipping field 102672883 because its id > 16777215
[874.815]    proto_decoder.cc:140 Skipping field 102672883 because its id > 16777215
[874.816]    proto_decoder.cc:140 Skipping field 102672883 because its id > 16777215
[874.816]    proto_decoder.cc:140 Skipping field 102672883 because its id > 16777215
[874.816]    proto_decoder.cc:140 Skipping field 102672883 because its id > 16777215
[874.816]    proto_decoder.cc:140 Skipping field 102672883 because its id > 16777215
Loading trace 0.43 MB (1.0 MB/s)
[889.711]      http_server.cc:271 [HTTP] POST /status [body=0B, origin="http://localhost:10000"]
[889.717]      http_server.cc:271 [HTTP] POST /status [body=0B, origin="http://localhost:10000"]
[897.966]      http_server.cc:271 [HTTP] POST /status [body=0B, origin="http://localhost:10000"]
[898.040]      http_server.cc:120 [HTTP] New connection
[898.040]      http_server.cc:271 [HTTP] GET /websocket/new [body=0B, origin="http://localhost:10000"]
[898.040]            httpd.cc:420 New TP instance 2 created via /websocket/new
[898.040]      http_server.cc:358 [HTTP] Handshaking WebSocket for /websocket/new
[898.040]      http_server.cc:360 > Upgrade: websocket
[898.040]      http_server.cc:360 > Connection: Upgrade
[898.040]      http_server.cc:360 > Sec-WebSocket-Accept: gFjJUl4pEdQGTl3zDglN884YlaI=
[899.398] ing_trace_parser.cc:108 proto trace detected
[899.552]    proto_decoder.cc:140 Skipping field 102672883 because its id > 16777215
[899.553]    proto_decoder.cc:140 Skipping field 102672883 because its id > 16777215
[899.553]    proto_decoder.cc:140 Skipping field 102672883 because its id > 16777215
[899.554]    proto_decoder.cc:140 Skipping field 102672883 because its id > 16777215
[899.554]    proto_decoder.cc:140 Skipping field 102672883 because its id > 16777215
[899.565]    proto_decoder.cc:140 Skipping field 102672883 because its id > 16777215
[899.565]    proto_decoder.cc:140 Skipping field 102672883 because its id > 16777215
[899.565]    proto_decoder.cc:140 Skipping field 102672883 because its id > 16777215
[899.565]    proto_decoder.cc:140 Skipping field 102672883 because its id > 16777215
[899.565]    proto_decoder.cc:140 Skipping field 102672883 because its id > 16777215
Loading trace 67.11 MB (2.1 MB/s))
[971.251]      http_server.cc:127 [HTTP] Client disconnected

------------------ BEGINNING OF CRASH ------------------
Signal: Segmentation fault
Fault addr: 0000000000000010

Backtrace:

#00  0x00007FED915B2DF0
     ./signal/../sysdeps/unix/sysv/linux/x86_64/libc_sigaction.c:0

#01  perfetto::base::UnixSocket::Send(void const*, unsigned long, int const*, unsigned long)
     /usr/local/google/home/sashwinbalaji/Desktop/perfetto/out/linux_clang_debug/../../src/base/unix_socket.cc:1195

#02  perfetto::base::UnixSocket::Send(void const*, unsigned long, int)
     ../../include/perfetto/ext/base/unix_socket.h:380

#03  perfetto::base::HttpServerConnection::SendWebsocketFrame(unsigned char, void const*, unsigned long)
     /usr/local/google/home/sashwinbalaji/Desktop/perfetto/out/linux_clang_debug/../../src/base/http/http_server.cc:592

#04  perfetto::base::HttpServerConnection::SendWebsocketMessage(void const*, unsigned long)
     /usr/local/google/home/sashwinbalaji/Desktop/perfetto/out/linux_clang_debug/../../src/base/http/http_server.cc:565

#05  perfetto::trace_processor::(anonymous namespace)::SendRpcChunk(perfetto::base::HttpServerConnection*, void const*, unsigned int)
     /usr/local/google/home/sashwinbalaji/Desktop/perfetto/out/linux_clang_debug/../../src/trace_processor/rpc/httpd.cc:222

#06  perfetto::trace_processor::(anonymous namespace)::Httpd::RpcThread::OnWebsocketMessage(perfetto::base::WebsocketMessage const&)::'lambda0'()::operator()() const::'lambda'(void const*, unsigned int)::operator()(void const*, unsigned int) const
     /usr/local/google/home/sashwinbalaji/Desktop/perfetto/out/linux_clang_debug/../../src/trace_processor/rpc/httpd.cc:165

#07  decltype(std::declval<perfetto::trace_processor::(anonymous namespace)::Httpd::RpcThread::OnWebsocketMessage(perfetto::base::WebsocketMessage const&)::'lambda0'()::operator()() const::'lambda'(void const*, unsigned int)&>()(std::declval<void const*>(), s
     ../../buildtools/libcxx/include/__type_traits/invoke.h:150

#08  void std::__Cr::__invoke_void_return_wrapper<void, true>::__call<perfetto::trace_processor::(anonymous namespace)::Httpd::RpcThread::OnWebsocketMessage(perfetto::base::WebsocketMessage const&)::'lambda0'()::operator()() const::'lambda'(void const*, unsig
     ../../buildtools/libcxx/include/__type_traits/invoke.h:226

#09  std::__Cr::__function::__default_alloc_func<perfetto::trace_processor::(anonymous namespace)::Httpd::RpcThread::OnWebsocketMessage(perfetto::base::WebsocketMessage const&)::'lambda0'()::operator()() const::'lambda'(void const*, unsigned int), void (void 
     ../../buildtools/libcxx/include/__functional/function.h:208

#0A  _ZNSt4__Cr10__function16__policy_invokerIFvPKvjEE11__call_implINS0_20__default_alloc_funcIZZN8perfetto15trace_processor12_GLOBAL__N_15Httpd9RpcThread18OnWebsocketMessageERKNS8_4base16WebsocketMessageEENKUlvE0_clEvEUlS3_jE_S4_EEEEvPKNS0_16__policy_storage
     ../../buildtools/libcxx/include/__functional/function.h:608

#0B  std::__Cr::__function::__policy_func<void (void const*, unsigned int)>::operator()(void const*&&, unsigned int&&) const
     ../../buildtools/libcxx/include/__functional/function.h:714

#0C  std::__Cr::function<void (void const*, unsigned int)>::operator()(void const*, unsigned int) const
     ../../buildtools/libcxx/include/__functional/function.h:987

#0D  perfetto::trace_processor::(anonymous namespace)::Response::Send(std::__Cr::function<void (void const*, unsigned int)>)
     /usr/local/google/home/sashwinbalaji/Desktop/perfetto/out/linux_clang_debug/../../src/trace_processor/rpc/rpc.cc:94

#0E  perfetto::trace_processor::Rpc::ParseRpcRequest(unsigned char const*, unsigned long)
     /usr/local/google/home/sashwinbalaji/Desktop/perfetto/out/linux_clang_debug/../../src/trace_processor/rpc/rpc.cc:255

#0F  perfetto::trace_processor::Rpc::OnRpcRequest(void const*, unsigned long)
     /usr/local/google/home/sashwinbalaji/Desktop/perfetto/out/linux_clang_debug/../../src/trace_processor/rpc/rpc.cc:127

#10  perfetto::trace_processor::(anonymous namespace)::Httpd::RpcThread::OnWebsocketMessage(perfetto::base::WebsocketMessage const&)::'lambda0'()::operator()() const
     /usr/local/google/home/sashwinbalaji/Desktop/perfetto/out/linux_clang_debug/../../src/trace_processor/rpc/httpd.cc:166

#11  decltype(std::declval<perfetto::trace_processor::(anonymous namespace)::Httpd::RpcThread::OnWebsocketMessage(perfetto::base::WebsocketMessage const&)::'lambda0'()&>()()) std::__Cr::__invoke<perfetto::trace_processor::(anonymous namespace)::Httpd::RpcThre
     ../../buildtools/libcxx/include/__type_traits/invoke.h:150

#12  void std::__Cr::__invoke_void_return_wrapper<void, true>::__call<perfetto::trace_processor::(anonymous namespace)::Httpd::RpcThread::OnWebsocketMessage(perfetto::base::WebsocketMessage const&)::'lambda0'()&>(perfetto::trace_processor::(anonymous namespac
     ../../buildtools/libcxx/include/__type_traits/invoke.h:226

#13  std::__Cr::__function::__default_alloc_func<perfetto::trace_processor::(anonymous namespace)::Httpd::RpcThread::OnWebsocketMessage(perfetto::base::WebsocketMessage const&)::'lambda0'(), void ()>::operator()()
     ../../buildtools/libcxx/include/__functional/function.h:208

#14  void std::__Cr::__function::__policy_invoker<void ()>::__call_impl<std::__Cr::__function::__default_alloc_func<perfetto::trace_processor::(anonymous namespace)::Httpd::RpcThread::OnWebsocketMessage(perfetto::base::WebsocketMessage const&)::'lambda0'(), v
     ../../buildtools/libcxx/include/__functional/function.h:608

#15  std::__Cr::__function::__policy_func<void ()>::operator()() const
     ../../buildtools/libcxx/include/__functional/function.h:714

#16  std::__Cr::function<void ()>::operator()() const
     ../../buildtools/libcxx/include/__functional/function.h:987

#17  perfetto::base::RunTaskWithWatchdogGuard(std::__Cr::function<void ()> const&)
     ../../include/perfetto/ext/base/watchdog.h:67

#18  perfetto::base::UnixTaskRunner::RunImmediateAndDelayedTask()
     /usr/local/google/home/sashwinbalaji/Desktop/perfetto/out/linux_clang_debug/../../src/base/unix_task_runner.cc:164

#19  perfetto::base::UnixTaskRunner::Run()
     /usr/local/google/home/sashwinbalaji/Desktop/perfetto/out/linux_clang_debug/../../src/base/unix_task_runner.cc:61

#1A  perfetto::trace_processor::(anonymous namespace)::Httpd::RpcThread::RpcThread()::'lambda'()::operator()() const
     /usr/local/google/home/sashwinbalaji/Desktop/perfetto/out/linux_clang_debug/../../src/trace_processor/rpc/httpd.cc:103

#1B  decltype(std::declval<perfetto::trace_processor::(anonymous namespace)::Httpd::RpcThread::RpcThread()::'lambda'()>()()) std::__Cr::__invoke<perfetto::trace_processor::(anonymous namespace)::Httpd::RpcThread::RpcThread()::'lambda'()>(perfetto::trace_proce
     ../../buildtools/libcxx/include/__type_traits/invoke.h:150

#1C  void std::__Cr::__thread_execute<std::__Cr::unique_ptr<std::__Cr::__thread_struct, std::__Cr::default_delete<std::__Cr::__thread_struct>>, perfetto::trace_processor::(anonymous namespace)::Httpd::RpcThread::RpcThread()::'lambda'()>(std::__Cr::tuple<std::
     ../../buildtools/libcxx/include/__thread/thread.h:194

#1D  void* std::__Cr::__thread_proxy<std::__Cr::tuple<std::__Cr::unique_ptr<std::__Cr::__thread_struct, std::__Cr::default_delete<std::__Cr::__thread_struct>>, perfetto::trace_processor::(anonymous namespace)::Httpd::RpcThread::RpcThread()::'lambda'()>>(void*
     ../../buildtools/libcxx/include/__thread/thread.h:202

#1E  start_thread
     ./nptl/pthread_create.c:448

#1F  __clone3
     ../sysdeps/unix/sysv/linux/x86_64/clone3.S:80

#20  0x0000000000000000
------------------ END OF CRASH ------------------
Segmentation fault (core dumped)
```

WangJialin547 committed:
solved segfault issue caused by sending rpc chunk to nonexistent connection

WangJialin547 committed:
added a /close http endpoint that kills unused tps

WangJialin547 committed:
changed modal id sorting from descending to ascending

WangJialin547 commented:
@sashwinbalaji Hi, all requested changes have been addressed. The crash was caused because the connection was closed when the trace is still being loaded and the rpc ended up sending a rpc chunk to a nonexistent connection, causing the crash. I added a check to ensure that the connection is still being registered by the httpd before sending the chunk. 

I also added the red X next to each row that, once clicked, sends a /close http request to the httpd and kills the respective tp. I also changed the order of the tps listed in the modal to ascending order. Here's what the new modal looks like: 

<img width="2560" height="1528" alt="image" src="https://github.com/user-attachments/assets/22629391-1b9d-4fe5-9892-2a2025a19a00" />

sashwinbalaji reviewed (changes requested):
Thanks the segfault is fixed. One issue with the introduction of "close" option.

If I close an attached session from other tab, the, the tp instance is removed but connection continues ? so that tab remains in a inconsistent state where we keep tryin to load but nothing works with error spam of

 httpd.cc:615 Websocket message from an un-associated connection.

It would be better if we aligned the flow and ensure on close proper cleanup occurs and on that tab the old ui of "Unable to connect to the device via websocket

trace_processor_shell --httpd is unreachable or crashed." is shown.

WangJialin547 committed:
kills the respective ws connection when /close request is sent, preventing error spam of 'unassociated connection'

WangJialin547 committed:
fix

WangJialin547 committed:
case fix

WangJialin547 committed:
fix

WangJialin547 committed:
change rpc_ to unique_ptr

WangJialin547 committed:
remove json dependency as no longer needed

WangJialin547 committed:
added detailed comment explaining the self-repeating cleanup task

WangJialin547 committed:
renamed has_active_tab to is_attached_ and added getters and setters

WangJialin547 committed:
format code

WangJialin547 committed:
fixed presubmit issues

WangJialin547 commented:
@sashwinbalaji hi, I have addressed all the suggested changes. I made sure to close the respective websocket connection after the red X has been clicked, so the respective tab will show the popup saying "unable to to connect to the device via websocket..." as intended. 

I also took care of all of the suggested changes, such as renaming has_existing_tab to is_attached_ and added getters and setters.

sashwinbalaji commented on src/trace_processor/trace_processor_shell.cc:
```diff
@@ -764,6 +765,8 @@ General purpose:
                                       executing some other commands (-q, -Q,
                                       --run-metrics, --summary).
+ --trace-processor-timeout-mins MINS  Specifies the timeout in minutes before the
+                                      trace processor thread to expire. Default is 120.
```
pls update this also.

sashwinbalaji commented on src/trace_processor/rpc/httpd.cc:
```diff
@@ -262,13 +614,79 @@ void Httpd::OnHttpRequest(const base::HttpRequest& req) {
+    const uint32_t instance_id = it->first;
+    uint64_t last_accessed = it->second->GetLastAccessedNs();
+
+    if (now - last_accessed > kInactivityNs) {
+      PERFETTO_ILOG(
+          "Cleaning up inactive RPC instance: %" PRIu32
+          " (inactive for %.1f minutes)",
+          instance_id,
+          static_cast<double>(now - last_accessed) / (kNanosecondPerMinute));
+      // Remove from conn_to_id_map as well
+      for (auto conn_it = conn_to_id_map.begin();
+           conn_it != conn_to_id_map.end();) {
+        if (conn_it->second == instance_id) {
+          conn_it = conn_to_id_map.erase(conn_it);
+        } else {
+          ++conn_it;
+        }
+      }
```
need to close connection here also ?

sashwinbalaji reviewed (changes requested):
Thanks most of flow looks as expected. Please fix the new comments.

Over to @stevegolton to review the UI part

stevegolton commented on ui/src/core/load_trace.ts:
```diff
@@ -220,6 +220,10 @@ async function loadTraceIntoEngine(
     engine,
   );
 
+  if (engine instanceof HttpRpcEngine) {
+    await sendTraceTitle(engine, traceSource);
+  }
+
```
Can this not be done inside the conditional above (just after `await engine.restoreInitialTables();`?)

WangJialin547 replied:
I am sparating conditional from the one above because they trigger on different conditions. the conditional above only gets triggered if the traceSource is 'HTTP_RPC', but we do not need to send the trace title in that case because a title has already been registered by  the TP.

The conditional we have here triggers if we are using the RPC, no matter what the trace source is. If the traceSource is FILE, URL, etc. we send the trace title for the TP to register. If the traceSource is HTTP_RPC then we do not do anything. For exmaple you open a TP on localhost:9001, then you use the "Open Trace" sidebar option to open another trace, in this case the traceSource is "FILE", and the engine is a HttpRpcEngine. If we combine it with the conditional above the trace title won't be sent correctly and the popup modal will show "Unnamed Trace", which can get confusing if we open multiple traces this way.

I will change the function name to maybeSendTraceTitle so it is more intuitive and reflects that the trace title does not always get sent

stevegolton commented on ui/src/core/load_trace.ts:
```diff
@@ -539,3 +543,27 @@ async function getTracingMetadataTimeBounds(engine: Engine): Promise<TimeSpan> {
   return new TimeSpan(startBound, endBound);
 }
+
+async function sendTraceTitle(engine: HttpRpcEngine, traceSource: TraceSource) {
```
Hang on, if engine is of type HttpRpcEngine, then surely our traceSource is going to always be 'HTTP_RPC' and this function will always do nothing? Or have I misunderstood something about how this works now?

WangJialin547 replied:
This function is for when you already have a tp running on 127.0.0.1:9001 and are opening a trace from a file, url, etc..
 
For example if you open a new trace file via the "Open trace file" sidebar option under RPC mode, the traceSource is 'FILE', the old version of the ui does not have a way to let the tp know what the trace name is, so the popup modal will show "Unnamed Trace" for all trace files opened that way.
 
If you open a file that is already existing in the tp, e.g. using trace_processor_shell then the traceSource is 'HTTP_RPC' and the tp has already registered the trace name so we don't need to resend the trace name, so we do nothing.

stevegolton commented on ui/src/frontend/index.ts:
```diff
@@ -92,6 +92,8 @@ function setupContentSecurityPolicy() {
     'http://127.0.0.1:9001', // For trace_processor_shell --httpd.
     'ws://127.0.0.1:9001', // Ditto, for the websocket RPC.
     'ws://127.0.0.1:9167', // For Web Device Proxy.
+    `ws://${window.location.hostname}:9001`,
+    `http://${window.location.hostname}:9001`,
```
Why do we need this?

stevegolton commented on ui/src/assets/preloaded_trace_dialog.scss:
```diff
@@ -0,0 +1,117 @@
+/* showDialogToUsePreloadedTrace.scss */
```
You have a lot of hard coded colors and styles in this stylesheet. For the sake of consistency and theme-ability, please use widgets where possible and use the theme colors in `theme_provider.scss`. Also, I think your LLM hallucinated a bunch of css variables that don't exist.

stevegolton commented on ui/src/trace_processor/http_rpc_engine.ts:
```diff
@@ -138,7 +187,7 @@ export class HttpRpcEngine extends EngineBase {
 
   static get hostAndPort() {
-    return `127.0.0.1:${HttpRpcEngine.rpcPort}`;
+    return `${window.location.hostname}:${HttpRpcEngine.rpcPort}`;
```
Why this change?

stevegolton commented on ui/src/frontend/rpc_http_dialog.ts:
```diff
@@ -304,36 +310,257 @@ enum PreloadedDialogResult {
+        ],
+      });
+
+      function redrawModal() {
+        try {
+          const mWithRedraw = m as {redraw?: () => void};
+          if (typeof mWithRedraw.redraw === 'function') {
+            mWithRedraw.redraw();
+          }
+        } catch (e) {}
+      }
```
You really shouldn't need this. For a start, m.redraw() already exists and is well defined in typescript. Also, the autoredraw system will automatically redraw after running event handlers registered via mithril

stevegolton reviewed (changes requested):
I've left a few preliminary comments. I'll need to test it in anger in order to provide more concrete feedback.

WangJialin547 committed:
changed trace_processor_shell comment

WangJialin547 committed:
also closes inactive WS connection in CleaUpInactiveInstances()

WangJialin547 committed:
moved sendTraceTitle next to restoreInitialtables

WangJialin547 committed:
removed window.location.hostname:9001 from rpcPolicy

WangJialin547 committed:
reverted change in http_rpc_engine.hostAndPort()

WangJialin547 committed:
removed redundant RedrawModal() function and replaced with m.redraw()

WangJialin547 committed:
removed unneccessary m.redraw

WangJialin547 committed:
Revert "moved sendTraceTitle next to restoreInitialtables"

This reverts commit edff7abeff6caa67bda2810b6a13b60078bbf5b3.

WangJialin547 committed:
used widgets and flexible style and pruned unneccessary styles

WangJialin547 committed:
format code

WangJialin547 commented:
@sashwinbalaji hi, both of the requested changes have been addressed

WangJialin547 commented:
@stevegolton hi, I have fixed the code according to the requested changes. I left some comments explaining my logic behind the send trace title changes: the old ui does not send the trace name to the TP so it will always use "Unnamed Trace". This can get confusing when we open more than one trace.

I also fixed rpc_http_dialog.ts to use widgets when possible, and pruned unneccessary css classes from preloaded_trace_dialog.scss. Now there are only 3 classes, representing the style of each row when it is selected, not selected, or unselectable because there is an active tab. All 3 classes use the colors in theme_provider.scss as requested

stevegolton reviewed (changes requested):
Hey, looks good in principle. I've left a few more some comments.

One thing I noticed is that when opening an existing instance from TP - the plugins start loading, then the trace is thrown away and the reloaded again in its entirety, so the plugins are loaded twice.

stevegolton committed:
ui: Allow any key-value pairs to be passed to (most) widgets (#3062)

Currently we restrict which attrs may be passed to widgets via their
attributes interfaces, most of which extend HTMLAttrs or one of its more
specific sub-interfaces. However, the number of possible HTML attributes
is large, and we don't want to have to hard code each possible type but
we also don't want to restrict users to only being able to use those
hard coded types.

This PR expands the HTMLAttrs type to allow arbitrary key value pairs on
top of the well known hard coded types. This provides type safety when
using the hard coded types, but also provides an escape hatch for when
users want to use new types, they just won't be type checked.

E.g.

```ts
// Before
m(Card, {className: 'foo', role: 'option'}) // Error because 'role' is not a member of HTMLAttrs.

// After
m(Card, {className: 'foo', role: 'option'}) // OK - No error, role's type just isn't checked (unknown), but other attributes such as className are.
```

WangJialin547 committed:
change trace_processsor_id to camelCase

WangJialin547 committed:
removed unneccessary undefined handling in http_rpc_engine.ts constructor

WangJialin547 committed:
utilize classNames utility

WangJialin547 committed:
improved generic css class names to specific ones, e.g. pf-rpc-http-dialog__row

WangJialin547 committed:
removed unneccessary Card as unknown as string after implementing #3062

WangJialin547 committed:
shows 'New Instance...' if there are no traces loaded for consistency

WangJialin547 committed:
removed unneccessary try catch clause

WangJialin547 committed:
added copyright header to preloaded_trace_dialog.scss

WangJialin547 committed:
deleted 'strong' inside the cards

WangJialin547 committed:
combined the x button with the stack

WangJialin547 committed:
removed keyboard ux

WangJialin547 committed:
removed side effects of useRpcWithPreloadedtrace and returns everything

WangJialin547 committed:
clicking the cards will now directly connect to the respective instance

WangJialin547 committed:
updated modal

WangJialin547 committed:
deleted preloaded_trace_dialog and uses default Card widget style

WangJialin547 committed:
removed the deleted file from perfetto.scss

WangJialin547 committed:
changed null to undefined

WangJialin547 committed:
format code

stevegolton reviewed (changes requested):
Thanks for making the changes - I added a few more comments.

In general I'm not super convinced about the way the rpc engine now needs a whole bunch more logic to obtain the traceProcessorId when requesting a new instance. It feels like the REST interface you have to look up the trace processor instances would be the ideal place where clients can request new instances - as it already acts as a higher level meta / management API on top of the low-level RPC websocket streaming API. Could we not add a POST endpoint here that allows us to create a new trace processor, returning the instance id, then we can just connect straight to it from the engine?

sashwinbalaji commented:
First of all, huge apologies from our end for the delay and multiple back and forth. We did a post-mortem on how to avoid these kinds of issues on external contributions and have come up with a process to handle them (will share soon in a separate discussion).

So, coming to this CL, we concluded that the rpc_engine should not handle any logic. The UI dialog you have built is great.

For the backend, we'd like to go with a different approach, somewhat similar to your previous implementation [change](https://github.com/google/perfetto/pull/2615/files/d626381005f4c60dba390d4304eb3cf6850c86e1). 
- Keep /status and its updated protos (RpcStatus & StatusResult).
- TP Selection:
    - If a new one is requested, the UI should create a new UUID and connect to /websocket/<new_uuid>
     - If an active TP is selected the UI should connect to its uuid as /websocket/<uuid>.
- The RPC layer should then keep a map from the UUID to the trace processor and use this to figure out which trace processor to use for the connection and if it’s an unknown one create a new TP instance.
- If no uuid is given (for back compat) it uses the "global" trace processor as it does today.

 Please keep the deletion/cleanup logic.

 Once again, thank you for your responsiveness and your contribution.

JialinWang28 commented:
@sashwinbalaji 

Hi there, thanks for checking in. I indeed am still actively working on this and my preference is that I still want to finish this PR myself.

Right now I’m in the middle of an unusually heavy academic period in college, with two major exams and multiple important deadlines next week (11/17 - 11/23), but I will be able to start making meaningful commits after that. 

If the project’s timeline can accommodate the delay, I’d really appreciate the chance to complete the remaining changes myself. If waiting causes issues, I understand, but my strong preference is to continue owning the PR.

Thank you again for your patience and understanding, and I will start working on this as soon as time allows.

sashwinbalaji commented:
Sure, take your time and please ensure it aligns with https://github.com/google/perfetto/pull/2615#issuecomment-3433786754. We are always happy to have external contributions.

WangJialin547 committed:
changed integer instance IDs to string UUIDs

WangJialin547 committed:
format code

WangJialin547 committed:
attempt to address back comp issues

WangJialin547 committed:
Merge remote-tracking branch 'upstream/main'

WangJialin547 committed:
pulled recent additions and made changes accordingly

WangJialin547 committed:
bug fix

WangJialin547 committed:
fix presubmit issues

WangJialin547 commented:
Hi @sashwinbalaji ,

Here's what I have made so far. I have addressed the issues stated in https://github.com/google/perfetto/pull/2615#issuecomment-3433786754 . I have 2 questions:

1. trace_processor_shell.cc allows the user to load a trace file in the command line; should I load it into the global rpc, or load it into a separate thread instead? 

In my current implementation, for backward comp I loaded it into the "global" rpc as requested in the comment. However due to the strcture of /status protos, this means that it can't:
a. display when it's last accessed; 
b. stop the user from opening it again in another tab if it's attached to an active tab already, which causes the old multitab sequence ID broken issue; 
c. be closed via the /close HTTP endpoint like the tps running on an RpcThread.

Should I leave it like this or implement some modifications?

2. I was away for a period of time and after I returned, I pulled everything new from the upstream branch, I resolved merge conflicts and updated neccessary things accordingly but met some presubmit errors regarding some modifications in Chromium while I didn't change anything. what should I do to resolve this?

--- Presubmit Errors Found ---
- Files under src/trace_processor/perfetto_sql/stdlib/chrome/, test/data/chrome/, test/trace_processor/diff_tests/stdlib/chrome/ are rolled from the Chromium repository by a Copybara service.
You should not modify these in the Perfetto repository, please make your changes in Chromium instead.
Affected files:
  - src/trace_processor/perfetto_sql/stdlib/chrome/OWNERS.github
  - src/trace_processor/perfetto_sql/stdlib/chrome/chrome_scrolls.sql
  - src/trace_processor/perfetto_sql/stdlib/chrome/event_latency.sql
  - src/trace_processor/perfetto_sql/stdlib/chrome/scroll_jank/scroll_jank_intervals.sql
  - src/trace_processor/perfetto_sql/stdlib/chrome/scroll_jank/scroll_jank_v3.sql
  - src/trace_processor/perfetto_sql/stdlib/chrome/scroll_jank/utils.sql
  - src/trace_processor/perfetto_sql/stdlib/chrome/scroll_jank_tagging.sql
  - test/trace_processor/diff_tests/stdlib/chrome/OWNERS.github
If you want to do a manual roll, you must specify CHROME_STDLIB_MANUAL_ROLL=<reason> in the CL description/commit message.

Thank you again for your time.

WangJialin547 committed:
loaded the trace_processor_shell instance onto a separate thread

WangJialin547 committed:
solved autoreconnection issue by keep track of uuids of closed connections

WangJialin547 committed:
fixed backcomp issues

LalitMaganti commented:
@WangJialin547 thank you for getting back to this patch. Most of the team (including @sashwinbalaji) have left for the Chrismas break so we will get back to reviewing this patch in the new year. Thanks again for your continued contribution and patience.

sashwinbalaji commented on protos/perfetto/trace_processor/trace_processor.proto:
```diff
@@ -257,9 +260,10 @@ message StatusArgs {}
-  // trace_processor_shell -D trace_file.pftrace .
+  // If present and not empty, a trace is already loaded already in one or more
+  // trace processors. This happens when using the HTTP+RPC mode and passing a
+  // trace file to the shell, via trace_processor_shell -D trace_file.pftrace .
```
nit: remove the .

sashwinbalaji commented on src/trace_processor/trace_processor_shell.cc:
```diff
@@ -761,6 +761,7 @@ struct CommandLineOptions {
   std::string metric_v1_names;
   std::string metric_v1_output;
   std::vector<std::string> raw_metric_v1_extensions;
+  size_t tp_timeout_mins = 24 * 60;  // 24 hours
```
please use uint32_t

sashwinbalaji commented on src/trace_processor/trace_processor_shell.cc:
```diff
@@ -2141,10 +2157,12 @@ base::Status TraceProcessorShell::Run(int argc, char** argv) {
     RunHttpRPCServer(
-        /*rpc=*/rpc,
+        /*preloaded_tp=*/rpc.TakeTraceProcessor(),
+        /*is_preloaded_eof-*/ !options.trace_file_path.empty(),
```
instead of 2 separate params (preloading tp and is_preloaded_eof) how about just continuing with passing rpc and exposing a getter in rpc to get eof_ ?

sashwinbalaji commented on src/trace_processor/rpc/httpd.cc:
```diff
@@ -147,14 +338,167 @@ void Httpd::OnHttpRequest(const base::HttpRequest& req) {
+        }
+        instance_uuid = parsed_uuid;
+        // check if this ID was previously closed
+        if (closed_conn_ids_.find(instance_uuid) != closed_conn_ids_.end()) {
```
please be consistent and use count everywhere if itr not needed

sashwinbalaji commented on src/trace_processor/rpc/httpd.cc:
```diff
@@ -62,12 +77,146 @@ class Httpd : public base::HttpRequestHandler {
+                         // causing a new thread to be reopened immediately.
+                         // when you make a new websocket handshake, check this
+                         // set to see if it was previously closed, and if so,
+                         // refuse to reopen it.
```
comment styling refer to others in the file.

sashwinbalaji reviewed (changes requested):
Thanks for your continued work on this.

Added some initial comments for now. To test it more please rebase your change and resolve the conflicts can be verified further.

LalitMaganti commented:
@WangJialin547 how is it going? What's the status of this PR? Are you still interested in completing it?

sashwinbalaji commented:
Hi @WangJialin547 

Catching up to see, if you are working on this PR ? If not, I can take it over and ensure you get the authorship for the change.

WangJialin547 commented:
Hi @sashwinbalaji , thanks for checking in. 

I’ve spent quite a bit of time working on this PR, but I’m currently too busy to finish it properly, so I think it would be the best if you take over. 

I really appreciate you offering to preserve my authorship for the change. I’m happy to share any notes to ensure smoother handoff, if needed. Thanks again!
